### PR TITLE
Taser Mode Refactor

### PR DESCRIPTION
**Pull request by realforest2001**
# About the pull request
Gave taser mode an action button. Uses the ID lock icon as it fits with the idea of using or not using the wanted database.
Also changed how it's handled on the back end, giving it a proc rather than replacing unique action with it.
Added an overlay on the taser itself that also shows what mode its using.

# Explain why it's good for the game
Makes it a bit more functional, also clearer to see what you're using.

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.
image
image
image

</details>

# Changelog

:cl:
refactor: Changing mode on the taser now uses an action button and unique proc rather than replacing unique action. Can still  be triggered with unique action.
imgadd: Added an overlay on the taser to denote what mode it's using.
/:cl:

### CODE REVIEW

File: code/modules/projectiles/guns/energy.dm [hunk 1]
```diff
@@ -200,7 +202,8 @@
 	charge_cost = 625 // approx 16 shots.
 	has_charge_meter = TRUE
 	charge_icon = "+taser"
-	var/precision = TRUE
+	actions_types = list(/datum/action/item_action/taser/change_mode)
+	var/mode = TASER_MODE_P
```
Comment (Zonespace27) [inline code/modules/projectiles/guns/energy.dm:206]: code doc

File: code/modules/projectiles/guns/energy.dm [hunk 2]
```diff
@@ -220,13 +223,64 @@
 			to_chat(user, SPAN_WARNING("You don't seem to know how to use [src]..."))
 			return FALSE
 
-/obj/item/weapon/gun/energy/taser/use_unique_action()
-	switch(precision)
-		if(TRUE)
-			precision = FALSE
-			to_chat(usr, SPAN_NOTICE("\The [src] is now set to Free mode."))
+/obj/item/weapon/gun/energy/taser/unique_action(mob/user)
+	change_mode(usr)
```
Comment (Zonespace27) [inline code/modules/projectiles/guns/energy.dm:227]: there's a perfectly good `user` argument already there

File: code/modules/projectiles/guns/energy.dm [hunk 3]
```diff
@@ -220,13 +223,64 @@
...
+
+/obj/item/weapon/gun/energy/taser/update_icon()
+	. = ..()
+	overlays += charge_icon + "_[mode]"
+
+/obj/item/weapon/gun/energy/taser/proc/change_mode(mob/user)
+	switch(mode)
+		if(TASER_MODE_P)
+			mode = TASER_MODE_F
+			to_chat(user, SPAN_NOTICE("\The [src] is now set to Free mode."))
```
Comment (Zonespace27) [inline code/modules/projectiles/guns/energy.dm:245]: \the is implicit

File: code/modules/projectiles/guns/energy.dm [hunk 4]
```diff
@@ -220,13 +223,64 @@
...
 			ammo = GLOB.ammo_list[/datum/ammo/energy/taser]
-		if(FALSE)
-			precision = TRUE
-			to_chat(usr, SPAN_NOTICE("\The [src] is now set to Precision mode."))
+		if(TASER_MODE_F)
+			mode = TASER_MODE_P
+			to_chat(user, SPAN_NOTICE("\The [src] is now set to Precision mode."))
```
Comment (Zonespace27) [inline code/modules/projectiles/guns/energy.dm:249]: ditto

File: code/modules/projectiles/guns/energy.dm [hunk 5]
```diff
@@ -220,13 +223,64 @@
...
 			ammo = GLOB.ammo_list[/datum/ammo/energy/taser/precise]
+	var/datum/action/item_action/taser/change_mode/action = locate(/datum/action/item_action/taser/change_mode) in actions
```
Comment (Zonespace27) [inline code/modules/projectiles/guns/energy.dm:251]: typecasting the var *might* make the locate type implicit? Would need to check this one

File: code/modules/projectiles/guns/energy.dm [hunk 6]
```diff
@@ -220,13 +223,64 @@
...
+	action.update_icon()
+	update_icon()
+	playsound(loc,'sound/machines/click.ogg', 25, 1)
+
+
+/datum/action/item_action/taser/action_activate()
```
Comment (Zonespace27) [inline code/modules/projectiles/guns/energy.dm:257]: this proc seemingly does nothing, as it returns nothing but null 100% of the time

File: code/modules/projectiles/guns/energy.dm [hunk 7]
```diff
@@ -220,13 +223,64 @@
...
+	var/obj/item/weapon/gun/energy/taser/taser = holder_item
+	if(!ishuman(owner))
+		return
+	var/mob/living/carbon/human/H = owner
```
Comment (Zonespace27) [inline code/modules/projectiles/guns/energy.dm:261]: one letter vars bad

File: code/modules/projectiles/guns/energy.dm [hunk 8]
```diff
@@ -220,13 +223,64 @@
...
+	button.name = name
+	button.overlays.Cut()
+	button.overlays += image('icons/mob/hud/actions.dmi', button, action_icon_state)
+
+/datum/action/item_action/taser/change_mode/action_activate()
+	. = ..()
+	var/obj/item/weapon/gun/energy/taser/taser = holder_item
+	taser.change_mode(usr)
+
+/datum/action/item_action/taser/change_mode/proc/update_icon()
```
Comment (Zonespace27) [inline code/modules/projectiles/guns/energy.dm:278]: code doc

File: code/modules/projectiles/guns/energy.dm [hunk 9]
```diff
@@ -220,13 +223,64 @@
...
+
+/datum/action/item_action/taser/action_activate()
+	var/obj/item/weapon/gun/energy/taser/taser = holder_item
+	if(!ishuman(owner))
+		return
+	var/mob/living/carbon/human/H = owner
+	if(H.is_mob_incapacitated() || taser.get_active_firearm(H, FALSE) != holder_item)
+		return
+
+/datum/action/item_action/taser/change_mode/New(Target, obj/item/holder)
```
Comment (Zonespace27) [inline code/modules/projectiles/guns/energy.dm:265]: don't capitalize args/vars

File: code/modules/projectiles/guns/energy.dm [hunk 10]
```diff
@@ -220,13 +223,64 @@
...
+		if(TASER_MODE_P)
+			. += SPAN_RED("It is set to precision mode, linked to the wanted database.")
+		if(TASER_MODE_F)
+			. += SPAN_GREEN("It is set to free mode, no longer linked to the wanted database.")
+
+/obj/item/weapon/gun/energy/taser/update_icon()
+	. = ..()
+	overlays += charge_icon + "_[mode]"
+
+/obj/item/weapon/gun/energy/taser/proc/change_mode(mob/user)
```
Comment (Zonespace27) [inline code/modules/projectiles/guns/energy.dm:241]: code doc

File: code/modules/projectiles/guns/energy.dm (hunk 9 above)
Comment (realforest2001) [reply code/modules/projectiles/guns/energy.dm:265]: I can probably change it without any consequence, but it's inherted from /datum/action

File: code/modules/projectiles/guns/energy.dm (hunk 6 above)
Comment (realforest2001) [reply code/modules/projectiles/guns/energy.dm:257]: Works as intended.

**Review by nauticall — CHANGES_REQUESTED** [review]
would suggest instead making the light a single led that switches between modes rather than its existing four-light combo. makes it sleeker and easier to distinguish.

suggested sprites are below, just copy+paste if you wanna use em!

image
image
image
image

**Review by harryob — APPROVED** [review]
code, needs sprite

**Pull request merged**